File: Functions/buildC50Phase.m
function c50 = buildC50Phase(dischSeg, chargeSegs, selectedVoltage, selectedCurrent, selectedTimeS)
% buildC50Phase - assemble a C/50 phase (discharge + matching charge) with
% a shared signed-Q axis integrated from the start of the discharge.
%
% The matching charge segment is the first C/50 charge segment whose start
% index is greater than the discharge end index. If no such charge segment
% exists, the discharge half is returned alone.
%
% Inputs:
%   dischSeg        - index vector of a C/50 discharge segment
%   chargeSegs      - cell array of C/50 charge segment index vectors
%   selectedVoltage - voltage array [V] for the selected range
%   selectedCurrent - current array [A] for the selected range
%   selectedTimeS   - seconds-since-start array [s] for the selected range
%
% Output:
%   c50 - struct with fields dischQ_Ah, dischV, chargeQ_Ah, chargeV
%
% Author: GitHub Copilot (for Feye Hoekstra)
% Date:   2026-07-28 (extracted from PlotCellSummary.m into a shared helper)

c50 = struct('dischQ_Ah',[], 'dischV',[], 'chargeQ_Ah',[], 'chargeV',[]);
if isempty(dischSeg); return; end
dischSeg = dischSeg(:);  % column

% Find the first charge segment that starts after the discharge end
dischEnd  = dischSeg(end);
matchedChg = [];
for ii = 1:numel(chargeSegs)
    if chargeSegs{ii}(1) > dischEnd
        matchedChg = chargeSegs{ii}(:);
        break
    end
end

if isempty(matchedChg)
    % Discharge-only fallback (no matching C/50 charge after this discharge)
    sTs = selectedTimeS(dischSeg); sTs = sTs - sTs(1);
    sI  = selectedCurrent(dischSeg);
    m   = ~isnan(sTs) & ~isnan(sI);
    Q   = NaN(size(sTs));
    if nnz(m) >= 2
        Q(m) = cumtrapz(sTs(m), sI(m)) / 3600;
    end
    c50.dischQ_Ah = Q;
    c50.dischV    = selectedVoltage(dischSeg);
    return
end

% Integrate over the FULL window from discharge start to charge end so the
% Q values at the start of the charge segment correctly include any
% (near-zero) current accumulated during the intervening rest.
fullRange = dischSeg(1) : matchedChg(end);
tFull = selectedTimeS(fullRange);
iFull = selectedCurrent(fullRange);

% NaN-safe trapz: linearly interpolate time across NaT gap markers, treat
% NaN-current as zero contribution. This is a no-op for clean segments.
iClean = iFull; iClean(isnan(iClean)) = 0;
tClean = tFull;
if any(isnan(tClean))
    idxAll = (1:numel(tClean))';
    good   = ~isnan(tClean);
    if nnz(good) >= 2
        tClean(~good) = interp1(idxAll(good), tClean(good), idxAll(~good), ...
            'linear', 'extrap');
    else
        tClean = idxAll;  % degenerate fallback
    end
end
cumQ_full_Ah = cumtrapz(tClean - tClean(1), iClean) / 3600;

% Index of each segment within fullRange (1-based, relative to dischSeg(1))
relDisch  = 1 : numel(dischSeg);
relCharge = (matchedChg(1) - dischSeg(1) + 1) : numel(fullRange);

c50.dischQ_Ah  = cumQ_full_Ah(relDisch);
c50.dischV     = selectedVoltage(dischSeg);
c50.chargeQ_Ah = cumQ_full_Ah(relCharge);
c50.chargeV    = selectedVoltage(matchedChg);
end

File: Functions/buildC50Phase.py
"""
buildC50Phase - assemble a C/50 phase (discharge + matching charge) with a
shared signed-Q axis integrated from the start of the discharge.

Python counterpart of Functions/buildC50Phase.m. Shared by
python_scripts/PlotCellSummary.py (todo #012).

Author: GitHub Copilot (for Robinson Medina / Feye Hoekstra)
Date: 2026-08-24
"""

import numpy as np

try:
    from scipy.integrate import cumulative_trapezoid as _cumtrapz
except ImportError:  # pragma: no cover
    from scipy.integrate import cumtrapz as _cumtrapz


def find_matching_charge_seg(disch_seg, charge_segs):
    """Return the first charge segment starting after the discharge ends, or None."""

    disch_end = disch_seg[-1]
    for seg in charge_segs:
        if seg[0] > disch_end:
            return seg
    return None


def build_c50_phase(disch_seg, charge_segs, selected_voltage, selected_current, selected_time_s):
    """
    Assemble a C/50 phase (discharge + matching charge) with a shared
    signed-Q axis integrated from the start of the discharge.

    Returns a dict with keys 'dischQ_Ah', 'dischV', 'chargeQ_Ah', 'chargeV'
    (charge* are empty arrays if no matching charge segment exists).
    """

    result = {'dischQ_Ah': np.array([]), 'dischV': np.array([]),
              'chargeQ_Ah': np.array([]), 'chargeV': np.array([])}
    if len(disch_seg) == 0:
        return result

    matched_chg = find_matching_charge_seg(disch_seg, charge_segs)

    if matched_chg is None:
        s_ts = selected_time_s[disch_seg]
        s_ts = s_ts - s_ts[0]
        s_i = selected_current[disch_seg]
        m = ~np.isnan(s_ts) & ~np.isnan(s_i)
        q = np.full(s_ts.shape, np.nan)
        if np.count_nonzero(m) >= 2:
            q[m] = _cumtrapz(s_i[m], s_ts[m], initial=0) / 3600
        result['dischQ_Ah'] = q
        result['dischV'] = selected_voltage[disch_seg]
        return result

    full_range = np.arange(disch_seg[0], matched_chg[-1] + 1)
    t_full = selected_time_s[full_range].astype(float)
    i_full = selected_current[full_range].astype(float)

    i_clean = np.nan_to_num(i_full, nan=0.0)
    t_clean = t_full.copy()
    if np.any(np.isnan(t_clean)):
        idx_all = np.arange(len(t_clean))
        good = ~np.isnan(t_clean)
        if np.count_nonzero(good) >= 2:
            t_clean[~good] = np.interp(idx_all[~good], idx_all[good], t_clean[good])
        else:
            t_clean = idx_all.astype(float)

    cum_q_full_ah = _cumtrapz(i_clean, t_clean - t_clean[0], initial=0) / 3600

    rel_disch = np.arange(0, len(disch_seg))
    rel_charge = np.arange(matched_chg[0] - disch_seg[0], len(full_range))

    result['dischQ_Ah'] = cum_q_full_ah[rel_disch]
    result['dischV'] = selected_voltage[disch_seg]
    result['chargeQ_Ah'] = cum_q_full_ah[rel_charge]
    result['chargeV'] = selected_voltage[matched_chg]
    return result


__all__ = ['build_c50_phase', 'find_matching_charge_seg']
